linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,180 @@
+# completion_port_linux requirements
+
+## Overview
+
+The completion_port_linux module handles the threading and control of the epoll socket system
+
+## Exposed API
+
+```C
+typedef struct COMPLETION_PORT_TAG* COMPLETION_PORT_HANDLE;
+
+#define COMPLETION_PORT_EPOLL_ACTION_VALUES \
+    COMPLETION_PORT_EPOLL_EPOLLRDHUP, \
+    COMPLETION_PORT_EPOLL_EPOLLIN, \
+    COMPLETION_PORT_EPOLL_EPOLLOUT
+
+MU_DEFINE_ENUM(COMPLETION_PORT_EPOLL_ACTION, COMPLETION_PORT_EPOLL_ACTION_VALUES)
+
+#define COMPLETION_PORT_EPOLL_RESULT_VALUES \
+    COMPLETION_PORT_RESULT_OK, \
+    COMPLETION_PORT_RESULT_ERROR, \
+    COMPLETION_PORT_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(COMPLETION_PORT_RESULT, COMPLETION_PORT_RESULT_VALUES)
+
+typedef void (*ON_COMPLETION_PORT_EVENT_COMPLETE)(void* context, COMPLETION_PORT_EPOLL_ACTION epoll_action, COMPLETION_PORT_RESULT result, int32_t amount_transfered);
+
+typedef struct EPOLL_BUFFER_DATA_TAG
+{
+    void* buffer;
+    uint32_t length;
+} EPOLL_BUFFER_DATA;
+
+typedef struct COMPLETION_PORT_INGRESS_DATA_TAG
+{
+    int epoll_op;
+    SOCKET_HANDLE socket;
+    ON_COMPLETION_PORT_EVENT_COMPLETE event_callback;
+    void* event_callback_ctx;
+    EPOLL_BUFFER_DATA buffer_data;
+} COMPLETION_PORT_INGRESS_DATA;
+
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, int, completion_port_post_queue, COMPLETION_PORT_HANDLE, completion_port, const COMPLETION_PORT_INGRESS_DATA*, ingress_data);
+MOCKABLE_FUNCTION(, void, completion_port_dequeue, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+### completion_port_create
+
+```C
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+```
+
+`completion_port_create` creates and initializes the completion port module
+
+`completion_port_create` shall create a completion port object.
+
+`completion_port_create` shall create the epoll module by calling `epoll_create`.
+
+`completion_port_create` shall create a thread to handle the epoll events.
+
+On success `completion_port_create` shall return the allocated `COMPLETION_PORT_HANDLE`.
+
+If there are any errors then `completion_port_create` shall fail and return `NULL`.
+
+### completion_port_dec_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_dec_ref` handles the decrement of the reference count and freeing the memory if necessary
+
+`completion_port_dec_ref` shall decrement the reference count for `completion_port`.
+
+If the reference count reaches 0, `completion_port_dec_ref` shall do the following:
+
+- increment the flag signaling that the threads can complete.
```

Comment:
>increment the flag [](http://example.com/codeflow?start=2&length=18)

This is weird, reading until here, this is the first reference to "flag". Also the only one, so I don't know what this "flag" is.

Also semantically talking, "flags" cannot really be "incremented". I always have this model where the "flag" is the little thingie attached to my mail box that I raise when there's mail inside, and the mailman lowers once he picks my mail up.

Reply:
Will change.